### ISSUE TITLE
[Bug] [Perf] Long non-overlapped cudaGraphLaunch in GLM-5.1 MTP target verify; propose pipelined two-graph decode replay

### ISSUE THREAD

**Issue by RunFMe** [sgl-project/sglang#29960]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

<img width="1970" height="715" alt="Image" src="https://github.com/user-attachments/assets/5c1cb344-7ea7-4a5d-8b83-13dcc263b35e" />

## Summary

When serving GLM-5.1 / GLM-5.1-derived models with EAGLE/MTP, target-verify decode CUDA graphs can become very large. In production profiling, `cudaGraphLaunch` for `step[TARGET_VERIFY ...]` takes multiple milliseconds on the scheduler main thread. This host-side launch time is not fully hidden by overlap scheduling and becomes visible latency.

The issue appears when replaying the full decode graph from:

    python/sglang/srt/model_executor/runner_backend/full_cuda_graph_backend.py(123): replay
    python/sglang/srt/model_executor/runner/decode_cuda_graph_runner.py(984): execute
    python/sglang/srt/model_executor/model_runner.py(3015): _forward_raw
    step[TARGET_VERIFY bs=...]
    python/sglang/srt/speculative/eagle_worker_v2.py(1447): verify

## Environment

- Model family: GLM-5.1 FP8 / fine-tuned GLM-5.1 FP8 with MTP heads
- Spec decoding: `--speculative-algorithm EAGLE`, `--speculative-num-steps 3`, `--speculative-eagle-topk 1`, `--speculative-num-draft-tokens 4`
- Decode CUDA graph backend: `full`
- Hardware: 8x NVIDIA H200, TP=8
- Trace metadata:
  - CUDA driver/runtime API: 13.0
  - CUPTI: 13.0.1
  - NCCL backend, world size 8

## Evidence from profiler

I am attaching a Perfetto screenshot showing a long `cudaGraphLaunch` on the scheduler main thread during target verify.

I am not attaching the full traces yet, but I extracted stats from the `cudaGraphLaunch` calls whose parent stack is target-verify work:

    torch/cuda/graphs.py(137): replay
    python/sglang/srt/model_executor/runner_backend/full_cuda_graph_backend.py(123): replay
    python/sglang/srt/model_executor/runner/decode_cuda_graph_runner.py(984): execute
    python/sglang/srt/model_executor/model_runner.py(3015): _forward_raw
    step[TARGET_VERIFY bs=...]

### TP0 profile

Target-verify `cudaGraphLaunch` stats:

- count: 18
- total host time: 116 ms
- p50: ~5.5 ms
- p90: ~8.8 ms
- max: 10.245 ms

The longest target-verify launches each submit ~3,475 GPU kernel nodes:

    10.245 ms cudaGraphLaunch -> 3,475 related GPU kernels, GPU span ~25.4 ms
     9.408 ms cudaGraphLaunch -> 3,475 related GPU kernels, GPU span ~20.4 ms
     8.496 ms cudaGraphLaunch -> 3,475 related GPU kernels, GPU span ~23.0 ms

### TP3 profile

Target-verify `cudaGraphLaunch` stats:

- count: 18
- total host time: approximately 150 ms
- p50: approximately 8 ms
- p90: approximately 12 ms
- max: 15.341 ms

Top launch stack:

    15.341 ms cudaGraphLaunch
      torch/cuda/graphs.py(137): replay
      python/sglang/srt/model_executor/runner_backend/full_cuda_graph_backend.py(123): replay
      python/sglang/srt/model_executor/runner/decode_cuda_graph_runner.py(984): execute
      python/sglang/srt/model_executor/model_runner.py(3015): _forward_raw
      step[TARGET_VERIFY bs=8]
      python/sglang/srt/speculative/eagle_worker_v2.py(1447): verify

This looks like host graph submission overhead for a very large full decode graph, not model compute itself and not a CUDA synchronization.

## Why this still matters with Spec V2 / overlap

There has been substantial work to make spec decoding overlappable:

- #11398 introduced beta EAGLE spec-overlap.
- #25879 routes `seq_lens` through `FutureMap` and removes `verify_done.wait()`.
- #25464 makes EAGLE / EAGLE3 / STANDALONE / MULTI_LAYER_EAGLE always use the V2 worker, with overlap scheduling driving it asynchronously.
- #25945 and #26380 further reduce scheduler-stream stalls and shared-buffer hazards.

However, the current profile still shows the scheduler main thread spending several milliseconds inside `cudaGraphLaunch` during target verify. Also, the event loop still has a result-processing sync path:

    python/sglang/srt/managers/scheduler.py(1549): pop_and_process
      -> process_batch_result(...)
         -> result.copy_done.synchronize()

So even in the fully overlapped V2 EAGLE worker, long host-side graph launches remain a problem: they can occupy the scheduler thread before the next useful CPU-side work can proceed, and any later `pop_and_process` synchronization can expose the launch bubble.

## Proposed fix: pipeline target-verify decode with two CUDA graphs

Instead of replaying target verify as one very large full decode graph, add an optional decode graph split:

    --decode-cuda-graph-split-nlayers N

or as CUDA graph config:

    {
      "decode": {
        "backend": "full",
        "split_nlayers": N
      }
    }

When enabled, capture two ordinary CUDA graphs per decode shape:

1. `graph_first`: metadata setup + embedding + first `N` transformer layers
2. `graph_second`: remaining layers + norm + logits

Replay:

    graph_first.replay()
    graph_second.replay()
    return graph_second_output

Both replays should happen on the same CUDA stream. No event or host sync is needed: stream ordering ensures graph 2 runs after graph 1. The benefit is that the CPU can launch `graph_second` while `graph_first` is already executing on the GPU. This should hide part of the second graph’s host submission cost behind useful GPU work.

For GLM-5.1-like 70B+ MoE models, the split point can be tuned. A starting sweep could be:

    split_nlayers = 4, 8, 16, 24, 32

The expected good region is where graph 1 has enough GPU work to cover most of graph 2’s host launch time, but graph 1 itself is still small enough to launch quickly.

## Implementation sketch

A minimal version could add a new decode backend, for example:

    LayerSplitFullCudaGraphBackend

It would be similar to `FullCudaGraphBackend`, but store:

    _graphs_first[shape_key]
    _graphs_second[shape_key]
    _outputs[shape_key]
    _bridge_tensors[shape_key]

The model forward needs a small split context:

    DecodeGraphSplitContext(
        phase="first" | "second",
        split_nlayers=N,
        hidden_states_bridge=...,
        residual_bridge=...,
        topk_indices_bridge=...,  # needed for DSA paths
    )

For GLM-5.1 DSA, this can initially be implemented in the DeepseekV2/GLM-DSA model loop:

- first phase: run layers `[start_layer, start_layer + N)`, copy `hidden_states`, `residual`, and DSA `topk_indices` into bridge tensors
- second phase: read bridge tensors, run `[start_layer + N, end_layer)`, then norm/logits normally

Important detail: metadata initialization should happen only in graph 1. Graph 2 should not rerun `attn_backend.init_forward_metadata_in_graph(...)`.

## Why not breakable CUDA graph?

Breakable CUDA graphs reduce the size of each launch but introduce eager CUDA ops between graph segments. In this case the target is different: keep the decode path as two fully captured CUDA graphs, with no eager CUDA work between them, and rely on launch pipelining on the same stream.

## Expected result

In the profiler, each target-verify step should show:

    cudaGraphLaunch(graph_first)   small host launch
    cudaGraphLaunch(graph_second)  host launch overlapping graph_first GPU execution

and no eager CUDA operations or `cudaStreamSynchronize` between the two launches.

The main metric is reduced exposed scheduler-thread time for target verify, especially for GLM-5.1 / GLM-DSA style models where the full decode graph contains thousands of graph nodes.

## Related upstream context

- https://github.com/sgl-project/sglang/issues/11762
- https://github.com/sgl-project/sglang/pull/11398
- https://github.com/sgl-project/sglang/pull/25879
- https://github.com/sgl-project/sglang/pull/25464
- https://github.com/sgl-project/sglang/pull/25945
- https://github.com/sgl-project/sglang/pull/26380

### Reproduction

```
# 1. Pick output dir for torch profiler traces
export SGLANG_TORCH_PROFILER_DIR=/tmp/sglang_glm51_mtp_profile
rm -rf "$SGLANG_TORCH_PROFILER_DIR"
mkdir -p "$SGLANG_TORCH_PROFILER_DIR"

# 2. Start server on 8 GPUs
python -m sglang.launch_server \
  --model-path zai-org/GLM-5.1-FP8 \
  --tp 8 \
  --tool-call-parser glm47 \
  --reasoning-parser glm45 \
  --speculative-algorithm EAGLE \
  --speculative-num-steps 3 \
  --speculative-eagle-topk 1 \
  --speculative-num-draft-tokens 4 \
  --cuda-graph-backend-decode full \
  --mem-fraction-static 0.85 \
  --host 0.0.0.0 \
  --port 30000

# 3. Send concurrent generation traffic and capture profiler trace.
# c=4 tends to create target-verify graphs with bs>1, where launch cost is visible.
python -m sglang.bench_serving \
  --backend sglang \
  --host 127.0.0.1 \
  --port 30000 \
  --model zai-org/GLM-5.1-FP8 \
  --dataset-name random \
  --random-input-len 2048 \
  --random-output-len 128 \
  --random-range-ratio 1.0 \
  --num-prompts 64 \
  --request-rate 4 \
  --profile
```

### Environment

I used recent main commit for tests but the problem should be present in older release versions too.

Python: 3.12.3
CUDA available: True
GPU: 8x NVIDIA H200
Compute capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 580.159.03
PyTorch: 2.11.0+cu130
sglang: 0.0.0.dev1+g828411e6f
sglang-kernel: 0.4.4
flashinfer_python: 0.6.12
flashinfer_cubin: 0.6.12
triton: 3.6.0
transformers: 5.8.1
torchao: 0.17.0+cu130
numpy: 2.3.5
xgrammar: 0.2.1
NCCL package in image: 2.28.3-1+cuda13.0
Container image: local/nova-gpu-main-breakable:no-sync-43696bb
Image SGLang build label: 828411e6f1adba2a6f8885119f4228439469951f
Image local revision label: 43696bbae8363f412b647a681120f01ff6c1c7c1

**Comment by RunFMe** [comment]
@Oasis-Git Hello! I think that this is relevant to the piece of code which you maintain, could you please give it a look? 
Can share my traces in .tgz if it helps

**Comment by Oasis-Git** [comment]
Thanks for report. Will take a look

**Comment by ormandj** [comment]
Measurements from DeepSeek-V4-Flash on 2x RTX PRO 6000 Blackwell GPUs (SM120, TP2, PCIe) with EAGLE/MTP2 (`num_steps=2`, top-k 1, verification tree width 3):

SGLang used main `1e10ec93` with the complete #28836, #29927, #30949, #32090, #32194, and #32220 PRs applied (tree `93323f7c`), Torch 2.13, and FlashInfer 0.6.14. The profile recorded CPU and CUDA activity with `with_stack=false` and `record_shapes=false`. The values below are TP0 medians from 12 steady-state C1 cycles.

| Metric | SGLang | vLLM |
|---|---:|---:|
| Verification cycle | 13.720 ms | 11.184 ms |
| CPU execution span | 7.294 ms | 5.074 ms |
| Non-graph CUDA launches per cycle | 47 | 28 |
| Target-verification graph nodes | 2,876 | 2,187 |
| Target-verification `cudaGraphLaunch` host time | 2.826 ms | 1.983 ms |
| Target-verification GPU span | 12.049 ms | 9.399 ms |
| Target-verification GPU union-busy time | 11.204 ms | 9.349 ms |
| Target-verification summed kernel time | 14.444 ms | 13.884 ms |

The two draft-graph GPU spans were 0.691 and 0.736 ms for SGLang and 0.646 and 0.616 ms for vLLM. The target-verification graph accounted for the cycle-time difference.

For the target-verification graph:

- Summed kernel time differed by 0.560 ms.
- `summed kernel time - union-busy time` differed by 1.295 ms, indicating 1.295 ms less cross-stream overlap in the SGLang trace.
- `GPU span - union-busy time` differed by 0.795 ms, indicating 0.795 ms more in-graph idle time in the SGLang trace.

The largest named summed-kernel differences were all-reduce at 0.330 ms, sparse attention at 0.152 ms, and unfused MHC at 0.087 ms. Together, these values do not account for the 2.650 ms target-graph span difference.

The serving runs had similar MTP acceptance: 66.11% for SGLang and 65.86% for vLLM.

#32060 addresses an oversized boolean fill that appears outside the target-verification graph once per cycle. It measured approximately 0.028 ms per cycle in this trace.

The vLLM trace used the same host, model, and MTP width but a different source and dependency lock, so the cross-engine values are not a matched dependency comparison. Both compressed traces are available if useful.

**Comment by RunFMe** [comment]
@Oasis-Git Hi! sorry for the repeated ping, could you take a look at say if it's worth pursuing and fixing?
